### PR TITLE
[mlir][Transforms] Add missing check in tosa::transpose::verify()

### LINKED ISSUE #99513: [mlir] Crash when using --pass-pipeline="builtin.module(func.func(tosa-to-linalg-named))"
I have the following MLIR program:
test.mlir:
```
func.func @func1() {
	%arg0 = tensor.empty() : tensor<3x4x5xi32>
	%1110 = arith.constant dense<[3, 0, 1]> : tensor<3xi32>
	%143 = tosa.transpose %arg0, %1110: (tensor<3x4x5xi32>, tensor<3xi32>) -> tensor<3x4x5xi32>
	return 
}
```
The above MLIR program will cause a crash when using the following command:
```
mlir-opt --pass-pipeline="builtin.module(func.func(tosa-to-linalg-named))" test.mlir
```
And the crash backtrace is:
```
mlir-opt: /data/tmp/v0718/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:257: const T& llvm::ArrayRef<T>::operator[](size_t) const [with T = mlir::OpFoldResult; size_t = long unsigned int]: Assertion `Index < Length && "Invalid index!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /data/tmp/v0718/llvm-project/build/bin/mlir-opt --pass-pipeline=builtin.module(func.func(tosa-to-linalg-named)) test.mlir
 #0 0x0000561de31ef8af llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x14f38af)
 #1 0x0000561de31ec924 SignalHandler(int) Signals.cpp:0:0
 #2 0x00007fc703d34420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #3 0x00007fc70380100b raise (/lib/x86_64-linux-gnu/libc.so.6+0x4300b)
 #4 0x00007fc7037e0859 abort (/lib/x86_64-linux-gnu/libc.so.6+0x22859)
 #5 0x00007fc7037e0729 (/lib/x86_64-linux-gnu/libc.so.6+0x22729)
 #6 0x00007fc7037f1fd6 (/lib/x86_64-linux-gnu/libc.so.6+0x33fd6)
 #7 0x0000561de4133b8f _ZN4llvm9to_vectorIRNS_14iterator_rangeINS_15mapped_iteratorINS_6detail15SafeIntIteratorIjLb0EEEZN4mlir16applyPermutationINS6_12OpFoldResultEEENS_11SmallVectorIT_XsrNS_42CalculateSmallVectorDefaultInlinedElementsISA_EE5valueEEENS_8ArrayRefISA_EENSE_IlEEEUllE_S8_EEEEEENS9_INSt12remove_constINSt16remove_referenceIDTdecl5begincl7declvalIRSA_EEEEE4typeEE4typeEXsrNSB_ISS_EE5valueEEEOSA_ (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x2437b8f)
 #8 0x0000561de5bab068 (anonymous namespace)::TransposeConverter::matchAndRewrite(mlir::tosa::TransposeOp, mlir::PatternRewriter&) const TosaToLinalgNamed.cpp:0:0
 #9 0x0000561de53ad64f mlir::detail::OpOrInterfaceRewritePatternBase<mlir::tosa::TransposeOp>::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x36b164f)
#10 0x0000561de939c848 mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<llvm::LogicalResult (mlir::Pattern const&)>) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x76a0848)
#11 0x0000561de6366c86 (anonymous namespace)::OperationLegalizer::legalize(mlir::Operation*, mlir::ConversionPatternRewriter&) DialectConversion.cpp:0:0
#12 0x0000561de636843e mlir::OperationConverter::convert(mlir::ConversionPatternRewriter&, mlir::Operation*) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x466c43e)
#13 0x0000561de636b13b mlir::OperationConverter::convertOperations(llvm::ArrayRef<mlir::Operation*>) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x466f13b)
#14 0x0000561de636b6ad mlir::applyFullConversion(mlir::Operation*, mlir::ConversionTarget const&, mlir::FrozenRewritePatternSet const&, mlir::ConversionConfig) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x466f6ad)
#15 0x0000561de5b594cd (anonymous namespace)::TosaToLinalgNamed::runOnOperation() TosaToLinalgNamedPass.cpp:0:0
#16 0x0000561de62c0861 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x45c4861)
#17 0x0000561de62c0cba mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x45c4cba)
#18 0x0000561de62c164c mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::'lambda'(mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo&)::operator()(mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo&) const Pass.cpp:0:0
#19 0x0000561de62bfdb5 mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x45c3db5)
#20 0x0000561de62c05e6 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x45c45e6)
#21 0x0000561de62c0cba mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x45c4cba)
#22 0x0000561de62c1474 mlir::PassManager::run(mlir::Operation*) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x45c5474)
#23 0x0000561de62b2a0b performActions(llvm::raw_ostream&, std::shared_ptr<llvm::SourceMgr> const&, mlir::MLIRContext*, mlir::MlirOptMainConfig const&) MlirOptMain.cpp:0:0
#24 0x0000561de62b33c4 processBuffer(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::MlirOptMainConfig const&, mlir::DialectRegistry&, llvm::ThreadPoolInterface*) MlirOptMain.cpp:0:0
#25 0x0000561de62b3574 llvm::LogicalResult llvm::function_ref<llvm::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&)::'lambda'(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>(long, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) MlirOptMain.cpp:0:0
#26 0x0000561de63bc2de mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<llvm::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, llvm::StringRef, llvm::StringRef) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x46c02de)
#27 0x0000561de62ab194 mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x45af194)
#28 0x0000561de62b36cc mlir::MlirOptMain(int, char**, llvm::StringRef, llvm::StringRef, mlir::DialectRegistry&) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x45b76cc)
#29 0x0000561de62b3b96 mlir::MlirOptMain(int, char**, llvm::StringRef, mlir::DialectRegistry&) (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x45b7b96)
#30 0x0000561de310ad6b main (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x140ed6b)
#31 0x00007fc7037e2083 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24083)
#32 0x0000561de31c2d9e _start (/data/tmp/v0718/llvm-project/build/bin/mlir-opt+0x14c6d9e)
Segmentation fault (core dumped)
```
My git version is 39bb244a16e59cb8f2080f96e9de599007762635.

### PR DESCRIPTION
**Pull request by DarshanRamakant**
The tosa::transpose::verify() should make sure
that the permutation numbers are within the size of 
the input array. Otherwise it will cause a cryptic array
out of bound assertion later.Fix #99513.

### CODE REVIEW

File: mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir
Comment (cxy-1993) [inline mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir:7]: nit: Could you please format your test as well?

File: mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir
Comment (cxy-1993) [inline mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir:6]: It seems strange to me: why tosa.transpose verify passed?

File: mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir
Comment (cxy-1993) [inline mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir:1]: I don't think we should put the test in this file; util and this conversion are not closely related.

File: mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir
Comment (cxy-1993) [inline mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir:4]: You've used very casual names for your functions and variables. This is not recommended.

File: mlir/include/mlir/Dialect/Utils/IndexingUtils.h
Comment (cxy-1993) [inline mlir/include/mlir/Dialect/Utils/IndexingUtils.h:207]: Top level question: Do we really need to do this check? The permutation itself represents its value range. If there's an error, it should have been caught before the call site. In what scenarios would we need this check?

**Review by cxy-1993 — CHANGES_REQUESTED** [review]
Thank you for your contribution. Personally, I don't think this assert is necessary. Perhaps we need to check the verify of other dialects.

File: mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir
Comment (DarshanRamakant) [reply mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir:7]: I will modify the testcase

File: mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir
Comment (DarshanRamakant) [reply mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir:6]: Yes you are right, there was no check in the tosa.transpose verifier. I have added a check there as well now.

File: mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir
Comment (DarshanRamakant) [reply mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir:4]: I will modify.

File: mlir/include/mlir/Dialect/Utils/IndexingUtils.h
Comment (DarshanRamakant) [reply mlir/include/mlir/Dialect/Utils/IndexingUtils.h:207]: I feel this check is needed. If you look at the original issue ([https://github.com/llvm/llvm-project/issues/99513]) there was a crash with cryptic assertion fail. This is happening because of the way permutation range is constructed :
`  auto permutationRange = llvm::map_range(
      llvm::seq<unsigned>(0, input.size()),
      [&](int64_t idx) -> T { return input[permutation[idx]]; });`

We want to make sure that we are not accessing out of boud while doing "input[permutation[idx]]"

File: mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir
Comment (DarshanRamakant) [reply mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir]: Removed this test, added this case in an existing test about invalid cases.

File: mlir/include/mlir/Dialect/Utils/IndexingUtils.h
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/Utils/IndexingUtils.h:205]: Would you be able to change the result of this to be a `FailureOr<SmallVector<T>>` That way, we can avoid the asserts

Comment (DarshanRamakant) [reply mlir/include/mlir/Dialect/Utils/IndexingUtils.h:205]: That change can be done. But that would lead to :
1) Changing the code of all the callers of "applyPermutation()" to take care of the failue path. 
2) Each caller can throw different error message on failure, which can lead to bad user experience. 

The original aim of this patch was to throw a meaningful error message. 
I can change the function signature in a new patch if all the reviewers agrees for that.

**Comment by DarshanRamakant** [comment]
@cxy-1993 @MaheshRavishankar  can you please merge this change if there are no other review comments ?

**Comment by DarshanRamakant** [comment]
Mistakenly closed the pull request, reopening it again.

File: mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalid.mlir
Comment (cxy-1993) [inline mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalid.mlir:47]: This test should be placed in mlir/test/Dialect/Tosa/invalid.mlir

Comment (cxy-1993) [inline mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalid.mlir:47]: You should pay more attention to the test format: reduce the use of tabs. Additionally, the variable names are too casual. %c1 doesn't match your value meaning at all, it is confusing. If you're unsure about naming variables, it's recommended to start with %0.

File: mlir/lib/Dialect/Tosa/IR/TosaOps.cpp
Comment (cxy-1993) [inline mlir/lib/Dialect/Tosa/IR/TosaOps.cpp:1120]: why no dynamic dims?

File: mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalid.mlir
Comment (DarshanRamakant) [reply mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalid.mlir:47]: Thanks for the feedback. I have used the test case as it is from the original issue. I will modify the names to use it from %0. 
Having said that many of the existing test cases ( even in the same folder) are not following naming convention that you have recommended.

File: mlir/lib/Dialect/Tosa/IR/TosaOps.cpp
Comment (DarshanRamakant) [reply mlir/lib/Dialect/Tosa/IR/TosaOps.cpp:1120]: Without that the existing behaviour will change. Couple of test cases in the shape-inference will give a different behaviour without that check
`https://github.com/llvm/llvm-project/blob/50a2b31800a554b35367a5f3bcebb640703a48e1/mlir/test/Dialect/Tosa/tosa-infer-shapes.mlir#L1398`
Above testcase will be having a different behaviour if we do it for dynamic shapes. So I tried to retain the original behaviour with that check.

File: mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalid.mlir
Comment (DarshanRamakant) [reply mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalid.mlir]: Updated

**Comment by DarshanRamakant** [comment]
@cxy-1993  @MaheshRavishankar @Mogball can you please merge this PR ?

**Comment by cxy-1993** [comment]
I maintain my original stance: when permutation is constant, verify doesn't require static shape. So I cant merge this for you. As for the test case you highlighted, I believe it would be more appropriate to modify it, such as by introducing perm as a parameter. And I don't think that's a good test, neither in terms of its style nor its test point design.

**Comment by DarshanRamakant** [comment]
> I maintain my original stance: when permutation is constant, verify doesn't require static shape. So I cant merge this for you. As for the test case you highlighted, I believe it would be more appropriate to modify it, such as by introducing perm as a parameter. And I don't think that's a good test, neither in terms of its style nor its test point design.

Aa suggested, I have removed the dynamic shape check, and modified the failing tests.

**Review by cxy-1993 — APPROVED** [review]
Thanks for iterating on this patch, LGTM.

**Comment by DarshanRamakant** [comment]
> Thanks for iterating on this patch, LGTM.

Thanks @cxy-1993, can you please merge the PR ?

**Comment by cxy-1993** [comment]
> > Thanks for iterating on this patch, LGTM.
> 
> Thanks @cxy-1993, can you please merge the PR ?

Yes. Could you please revise the commit message and PR description before merging? The information in them no longer aligns with the current code.

**Comment by DarshanRamakant** [comment]
> > > Thanks for iterating on this patch, LGTM.
> > 
> > 
> > Thanks @cxy-1993, can you please merge the PR ?
> 
> Yes. Could you please revise the commit message and PR description before merging? The information in them no longer aligns with the current code.

Done

File: mlir/test/Dialect/Tosa/invalid.mlir
Comment (cxy-1993) [inline mlir/test/Dialect/Tosa/invalid.mlir:425]: Do not use tab

**Review by cxy-1993 — COMMENTED** [review]
I've found another style issue... There's a problem with the CI in the LLVM, this patch isn't the cause of the current CI failure. Please fix this style issue, and then I'll help you merge the code.

File: mlir/test/Dialect/Tosa/invalid.mlir
Comment (DarshanRamakant) [reply mlir/test/Dialect/Tosa/invalid.mlir]: Strangely it was no visible in my local vs code. Fixed the issue now.
Yes, the latest CI failures not related, it is happening on all the patches.

**Pull request merged**
